File: src/Data_Structures/Binary_Tree/binary_tree_test.py
from binary_tree import BinaryTree


binary_tree = BinaryTree(5)

list = [3, 6, 1, 7, 8, 4, 10, 2, 9]
for value in list:
    binary_tree.insert(value)

    #               5
    #          3         6
    #       1     4         7
    #         2                8
    #                            10
    #                          9


def test_createBinaryTree():
    assert binary_tree.value == 5
    assert binary_tree.left.value == 3
    assert binary_tree.right.value == 6
    assert binary_tree.left.left.value == 1
    assert binary_tree.right.right.value == 7
    assert binary_tree.right.right.right.value == 8
    assert binary_tree.left.right.value == 4
    assert binary_tree.right.right.right.right.value == 10
    assert binary_tree.left.left.right.value == 2
    assert binary_tree.right.right.right.right.left.value == 9


def test_contains():
    assert binary_tree.contains(2) == True
    assert binary_tree.contains(9) == True
    assert binary_tree.contains(0) == False
    assert binary_tree.contains(11) == False


def test_depth_first_traversal():
    _pre = []
    _in = []
    _post = []

    def preCallback(value):
        _pre.append(value)

    def inCallback(value):
        _in.append(value)

    def postCallback(value):
        _post.append(value)

    binary_tree.depth_first_traversal("pre", preCallback)
    binary_tree.depth_first_traversal("in", inCallback)
    binary_tree.depth_first_traversal("post", postCallback)
    assert _pre == [5, 3, 1, 2, 4, 6, 7, 8, 10, 9]
    assert _in == [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
    assert _post == [2, 1, 4, 3, 9, 10, 8, 7, 6, 5]


def test_breadth_first_traversal():
    result = []

    def callback(value):
        result.append(value)

    binary_tree.breadth_first_traversal(callback)
    assert result == [5, 3, 6, 1, 4, 7, 2, 8, 10, 9]


def test_getMinValue():
    assert binary_tree.get_min_value() == 1


def test_get_max_value():
    assert binary_tree.get_max_value() == 10


def test_customBinaryTree():
    customBinaryTree = BinaryTree("a")
    b = customBinaryTree.insert_left("b")
    c = customBinaryTree.insert_right("c")
    d = b.insert_left("d")
    e = b.insert_right("e")
    f = c.insert_left("f")
    g = c.insert_right("g")
    h = d.insert_left("h")
    i = d.insert_right("i")

    #                  a
    #            b           c
    #         d     e     f     g
    #       h   i

    assert customBinaryTree.value == "a"
    assert customBinaryTree.left.value == "b"
    assert customBinaryTree.left.left.value == "d"
    assert customBinaryTree.left.right.value == "e"
    assert customBinaryTree.left.left.left.value == "h"
    assert customBinaryTree.left.left.right.value == "i"
    assert customBinaryTree.right.value == "c"
    assert customBinaryTree.right.left.value == "f"
    assert customBinaryTree.right.right.value == "g"

File: src/Data_Structures/Binary_Tree/binary-tree.test.ts
//@ts-nocheck

import { BinaryTree } from "./binary-tree";

describe("Binary Tree", () => {
  const binaryTree = new BinaryTree(5);

  beforeAll(() => {
    for (const value of [3, 6, 1, 7, 8, 4, 10, 2, 9]) {
      binaryTree.insert(value);
    }

    //               5
    //          3         6
    //       1     4         7
    //         2                8
    //                            10
    //                          9

    binaryTree.print(); //?
    binaryTree.print("pre"); //?
    binaryTree.print("post"); //?
  });

  test("Should create a binary tree", () => {
    expect(binaryTree.value).toBe(5);
    expect(binaryTree.left.value).toBe(3);
    expect(binaryTree.right.value).toBe(6);
    expect(binaryTree.left.left.value).toBe(1);
    expect(binaryTree.right.right.value).toBe(7);
    expect(binaryTree.right.right.right.value).toBe(8);
    expect(binaryTree.left.right.value).toBe(4);
    expect(binaryTree.right.right.right.right.value).toBe(10);
    expect(binaryTree.left.left.right.value).toBe(2);
    expect(binaryTree.right.right.right.right.left.value).toBe(9);
  });

  test("Should implement contains", () => {
    expect(binaryTree.contains(2)).toBeTruthy();
    expect(binaryTree.contains(9)).toBeTruthy();
    expect(binaryTree.contains(0)).toBeFalsy();
    expect(binaryTree.contains(11)).toBeFalsy();
  });

  test("Should implement depthFirstTraversal", () => {
    const _pre = [];
    const _in = [];
    const _post = [];
    binaryTree.depthFirstTraversal("pre", (value) => _pre.push(value));
    binaryTree.depthFirstTraversal("in", (value) => _in.push(value));
    binaryTree.depthFirstTraversal("post", (value) => _post.push(value));
    expect(_pre).toEqual([5, 3, 1, 2, 4, 6, 7, 8, 10, 9]);
    expect(_in).toEqual([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
    expect(_post).toEqual([2, 1, 4, 3, 9, 10, 8, 7, 6, 5]);
  });

  test("Should implement breadthFirstTraversal", () => {
    const result = [];
    binaryTree.breadthFirstTraversal((value) => result.push(value));
    expect(result).toEqual([5, 3, 6, 1, 4, 7, 2, 8, 10, 9]);
  });

  test("Should implement getMinValue", () => {
    expect(binaryTree.getMinValue()).toBe(1);
  });

  test("Should implement getMaxValue", () => {
    expect(binaryTree.getMaxValue()).toBe(10);
  });

  test("Should create a custom binary tree", () => {
    const customBinaryTree = new BinaryTree("a");
    const b = customBinaryTree.insertLeft("b");
    const c = customBinaryTree.insertRight("c");
    const d = b.insertLeft("d");
    const e = b.insertRight("e");
    const f = c.insertLeft("f");
    const g = c.insertRight("g");
    const h = d.insertLeft("h");
    const i = d.insertRight("i");

    //                  a
    //            b           c
    //         d     e     f     g
    //       h   i

    customBinaryTree.print(); //?
    customBinaryTree.print("pre"); //?
    customBinaryTree.print("post"); //?

    expect(customBinaryTree.value).toBe("a");
    expect(customBinaryTree.left.value).toBe("b");
    expect(customBinaryTree.left.left.value).toBe("d");
    expect(customBinaryTree.left.right.value).toBe("e");
    expect(customBinaryTree.left.left.left.value).toBe("h");
    expect(customBinaryTree.left.left.right.value).toBe("i");
    expect(customBinaryTree.right.value).toBe("c");
    expect(customBinaryTree.right.left.value).toBe("f");
    expect(customBinaryTree.right.right.value).toBe("g");
  });
});
